cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java

```diff
@@ -436,6 +456,80 @@ public void testArtifacts() throws Exception {
     Assert.assertEquals(ImmutableSet.of(), authorizer.listPrivileges(ALICE));
   }
 
+  @Test
+  public void testPrograms() throws Exception {
+    Authorizer authorizer = getAuthorizer();
+    authorizer.grant(instance, ALICE, ImmutableSet.of(Action.ADMIN));
+    getNamespaceAdmin().create(AUTH_NAMESPACE_META);
+    Assert.assertEquals(
+      ImmutableSet.of(new Privilege(instance, Action.ADMIN), new Privilege(AUTH_NAMESPACE, Action.ALL)),
+      authorizer.listPrivileges(ALICE)
+    );
+    final ApplicationManager dummyAppManager = deployApplication(AUTH_NAMESPACE.toId(), DummyApp.class);
+    ArtifactSummary dummyArtifactSummary = dummyAppManager.getInfo().getArtifact();
+    NamespacedArtifactId dummyArtifact = AUTH_NAMESPACE.artifact(dummyArtifactSummary.getName(),
+                                                                 dummyArtifactSummary.getVersion());
+    ApplicationId appId = AUTH_NAMESPACE.app(DummyApp.class.getSimpleName());
+    final ProgramId serviceId = appId.service(DummyApp.Greeting.SERVICE_NAME);
+    Assert.assertEquals(
+      ImmutableSet.of(
+        new Privilege(instance, Action.ADMIN),
+        new Privilege(AUTH_NAMESPACE, Action.ALL),
+        new Privilege(dummyArtifact, Action.ALL),
+        new Privilege(appId, Action.ALL),
+        new Privilege(serviceId, Action.ALL)
+      ),
+      authorizer.listPrivileges(ALICE)
+    );
+    // alice should be able to start and stop programs in the app she deployed
+    dummyAppManager.startProgram(serviceId.toId());
+    Tasks.waitFor(true, new Callable<Boolean>() {
+      @Override
+      public Boolean call() throws Exception {
+        return dummyAppManager.isRunning(serviceId.toId());
+      }
+    }, 5, TimeUnit.SECONDS);
+    dummyAppManager.stopProgram(serviceId.toId());
+    Tasks.waitFor(false, new Callable<Boolean>() {
+      @Override
+      public Boolean call() throws Exception {
+        return dummyAppManager.isRunning(serviceId.toId());
+      }
+    }, 5, TimeUnit.SECONDS);
+    // Bob should not be able to start programs in dummy app because he does not have privileges on it
+    SecurityRequestContext.setUserId(BOB.getName());
+    try {
+      dummyAppManager.startProgram(serviceId.toId());
+    } catch (RuntimeException e) {
+      //noinspection ThrowableResultOfMethodCallIgnored
+      Assert.assertTrue(Throwables.getRootCause(e) instanceof UnauthorizedException);
+    }
+    // can't verify running programs in this case, because DefaultApplicationManager maintains an in-memory map of
+    // running processes that does not use ApplicationLifecycleService to get the runtime status. So no matter if the
+    // start/stop call succeeds or fails, it updates its running state in the in-memory map.
+    try {
+      dummyAppManager.stopProgram(serviceId.toId());
```

Comment:
cannot add this verification because of https://issues.cask.co/browse/CDAP-5452. Will come back to it after fixing that bug.